ArduPlane/mode_autoland.cpp

```diff
@@ -0,0 +1,114 @@
+#include "mode.h"
+#include "Plane.h"
+#include <GCS_MAVLink/GCS.h>
+
+/*
+  mode AutoLand parameters
+ */
+const AP_Param::GroupInfo ModeAutoLand::var_info[] = {
+    // @Param: WP_ALT
+    // @DisplayName: Final approach WP altitude
+    // @Description: This is the target altitude above HOME for final approach waypoint
+    // @Range: 0 200
+    // @Increment: 1
+    // @Units: m
+    // @User: Standard
+    AP_GROUPINFO("WP_ALT", 1, ModeAutoLand, final_wp_alt, 55),
+
+    // @Param: WP_DIST
+    // @DisplayName: Final approach WP distance
+    // @Description: This is the distance from Home that the final approach waypoint is set. The waypoint point will be in the opposite direction of takeoff (the direction the plane is facing when the plane sets its takeoff heading)
+    // @Range: 0 700
+    // @Increment: 1
+    // @Units: m
+    // @User: Standard
+    AP_GROUPINFO("WP_DIST", 2, ModeAutoLand, final_wp_dist, 400),
+
+    // @Param: DIR_OFF
+    // @DisplayName: Landing direction offset from takeoff
+    // @Description: The captured takeoff direction (at arming,if TKOFF_OPTION bit1 is set, or after ground course is established in autotakeoffs)is offset by this amount to create a different landing direction and approach.
+    // @Range: -360 360
+    // @Increment: 1
+    // @Units: deg
+    // @User: Standard
+    AP_GROUPINFO("DIR_OFF", 3, ModeAutoLand, landing_dir_off, 0),
+
+    AP_GROUPEND
+};
+
+ModeAutoLand::ModeAutoLand() :
+    Mode()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+bool ModeAutoLand::_enter()
+{             
+    //must be flying to enter
+    if (!plane.is_flying()) { 
+        gcs().send_text(MAV_SEVERITY_WARNING, "Must already be flying!"); 
+        return false;
+    }
+    
+    //takeoff direction must be set and must not be a quadplane, otherwise since flying switch to RTL so this can be used as FS action
+#if HAL_QUADPLANE_ENABLED
+    if (quadplane.available()) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "AutoLand is fixed wing only mode"); 
```

Comment:
We should permit the use of this mode if the Q_OPTION bit is set to allow fixed-wing landings.

Reply:
fine by me and easy to do, but @tridge was the one who asked me to prevent Qplanes from using this

Reply:
He reminded me the Q_OPTION bit existed :-)

Reply:
@peterbarker thanks for the flight testing! added the ability to use this mode in QPlane if option is set....also landing auto-flaps WAS working, although incorrectly: they were activated immediately upon mode entry while flying to final waypoint...the last commit (unsquashed so changes are easily reviewed) fixes this....SITL tested for flaps
my namespace mojo is weak and still get a compile error on the Qoption addition even adding the QPlane friend class...need help fixing that